framework/encode/dx12_state_writer.cpp

```diff
@@ -747,7 +747,9 @@ void Dx12StateWriter::WriteResourceSnapshots(
                 auto     device_info   = device_wrapper->GetObjectInfo();
 
                 const double max_cpu_mem_usage = 7.0 / 8.0;
-                if (!graphics::dx12::IsMemoryAvailable(size_in_bytes, device_info.get()->adapter3, max_cpu_mem_usage))
+
+                const bool isUma = graphics::dx12::IsUma(*device_wrapper->GetWrappedObjectAs<ID3D12Device>());
```

Comment:
Can the call to `IsUma` be moved outside the inner loop, or its result be saved for each device? This would avoid making repeated API calls to `ID3D12Device::CheckFeatureSupport` for each device.

Reply:
Existing code gets the device wrapper from the snapshot which the loop yields. I'd have to cache at least the isUma bool as a device wrapper member. I did want to do that but thought it might be considered too intrusive. If you and Brad like that approach, I'm happy to go that way too.

Question, should the cached var go in the device wrapper or the reference counted `ID3D12DeviceInfo` that the wrapper points at? The latter seems to fit the pattern of the existing code but bouncing around in memory two hops to look up a cached bool seems _out of keeping with the aim of going fast_. In this case in the state writer we do touch the info object for its other contents anyway. Maybe we will always do so in other cases where we'd want the cached flag.

Reply:
I've implemented caching on encode and decode sides. This captures with trimming and the capture replays without logging the error message from issue #954 (`[gfxrecon] ERROR - Detected adapter memory oversubscription`).

If you two are happy and CI passes, I'll call that done.